src/common/column_matrix.h

```diff
@@ -286,22 +297,18 @@ class ColumnMatrix {
         }
       };
 
-      const xgboost::Entry* data_ptr = page.data.HostVector().data();
-      const std::vector<bst_row_t>& offset_vec = page.offset.HostVector();
       const size_t batch_size = gmat.Size();
-      CHECK_LT(batch_size, offset_vec.size());
+      size_t k{0};
       for (size_t rid = 0; rid < batch_size; ++rid) {
-        const size_t ibegin = gmat.row_ptr[rid];
-        const size_t iend = gmat.row_ptr[rid + 1];
-        const size_t size = offset_vec[rid + 1] - offset_vec[rid];
-        SparsePage::Inst inst = {data_ptr + offset_vec[rid], size};
-
-        CHECK_EQ(ibegin + inst.size(), iend);
-        size_t j = 0;
-        for (size_t i = ibegin; i < iend; ++i, ++j) {
-          const uint32_t bin_id = row_index[i];
-          auto fid = inst[j].index;
-          get_bin_idx(bin_id, rid, fid);
+        auto line = batch.GetLine(rid);
+        for (size_t i = 0; i < line.Size(); ++i) {
+          auto coo = line.GetElement(i);
+          if (data::IsValidFunctor {missing}(coo)) {
+            auto fid = coo.column_idx;
+            const uint32_t bin_id = row_index[k];
```

Comment:
What's the performance implication of abstracting the element access? Is the overhead acceptable?

Reply:
Also,  wouldn't it be inefficient to create the functor in a tight-loop? Let's create the functor instance outside the loop.

Reply:
> Is the overhead acceptable?

It's not the bottleneck. The major issue with this function is the `missing_flags_` that's a boolean vector, which is not thread-safe. https://github.com/dmlc/xgboost/pull/7208 is likely to rewrite many of these anyway, right now I just want to make sure some of the features are merged before over optimization.

Reply:
> right now I just want to make sure some of the features are merged before optimization.

I'm not against optimization (and like them), just trying to show a big picture of what we need in these data structures before we go nuts with optimization.